### ISSUE TITLE
onnx-community/OpenMed-PII-SuperClinical: q8 silently returns near-chance logits; fp16 and q4f16 fail to load

### ISSUE THREAD

**Issue by matthewchung74** [huggingface/transformers.js#1749]
### Summary

Three of the five ONNX variants published for [`onnx-community/OpenMed-PII-SuperClinical-Base-184M-v1-ONNX`](https://huggingface.co/onnx-community/OpenMed-PII-SuperClinical-Base-184M-v1-ONNX) are unusable. Two fail loudly at session creation; **one loads fine and silently returns near-chance predictions**, which is the reason I'm filing — a consumer of `q8` gets a pipeline that appears healthy while detecting essentially nothing.

| Variant | File | Size | Result |
|---|---|---|---|
| `fp32` | `model.onnx` | 704 MB | ✅ works — max score 1.00, correct labels |
| `q4` | `model_q4.onnx` | 496 MB | ✅ works — max score 0.999, correct labels |
| `q8` | `model_quantized.onnx` | 232 MB | ❌ **loads, but predictions are noise** |
| `fp16` | `model_fp16.onnx` | 369 MB | ❌ fails to load (float16/float32 cast mismatch) |
| `q4f16` | `model_q4f16.onnx` | 275 MB | ❌ fails to load (same class of error) |

### Environment

- `@huggingface/transformers` 4.2.0
- `onnxruntime-node` 1.24.3
- Node v23.6.1, macOS (Darwin arm64)

All three failures reproduced in Node. The `q8` behaviour and the `q4` fix were additionally confirmed in-browser with the same library version from jsDelivr; `fp16` and `q4f16` were not retested in a browser.

### Reproduction

```js
import { pipeline } from "@huggingface/transformers";

const MODEL = "onnx-community/OpenMed-PII-SuperClinical-Base-184M-v1-ONNX";
const text = "Patient: Jane Q. Testpatient DOB: 03/14/1985 MRN: TESTMRN-424242 seen by Dr. Alice Smith at 123 Canary Lane.";

for (const dtype of ["q8", "q4", "fp16", "q4f16", "fp32"]) {
  try {
    const p = await pipeline("token-classification", MODEL, { dtype });
    const out = await p(text, { ignore_labels: ["O"] });
    console.log(dtype, out.length, "entities, max score",
      out.length ? Math.max(...out.map(o => o.score)).toFixed(3) : "n/a");
  } catch (e) {
    console.log(dtype, "FAILED:", e.message);
  }
}
```

### 1. `q8` — loads, but output is near-chance and mislabelled

```
[ { entity: "I-phone_number",               score: 0.042, word: "Jane"  },
  { entity: "B-certificate_license_number", score: 0.033, word: "Alice" } ]
```

Two spans, both mislabelled, both at roughly chance — `Jane` as a phone number, `Alice` as a licence number. No error is raised.

Same input and code with `dtype: "q4"`:

```
11 entities, max score 0.999
Jane        -> B-first_name    @ 1.00
03/14/1985  -> date_of_birth   @ 0.99
Alice, Smith, Robert, Chen  all detected as names
```

The near-uniform softmax under `q8` suggests quantization damaged the classifier head rather than a tokenizer or config problem.

### 2. `fp16` — fails to load

```
Load model from .../onnx/model_fp16.onnx failed:
Type Error: Type (tensor(float16)) of output arg (/deberta/embeddings/Cast_output_0)
of node (/deberta/embeddings/Cast) does not match expected type (tensor(float)).
```

The `Cast` node at the DeBERTa embeddings emits `float16` where the consumer expects `float32`, so the graph fails type-checking at session creation. The file downloads completely (369 MB), so this isn't a truncated artifact.

### 3. `q4f16` — fails to load

Same failure mode as `fp16` (mixed-precision cast mismatch). File is 275 MB and downloads completely.

### Why this matters for web deployment

`q8` (232 MB) and `fp16` (369 MB) are the two sizes most attractive for browser use. With both unusable, the smallest working option is `q4` at 496 MB — more than double the broken `q8`. We shipped `q4` for an on-device PII redaction feature and it works well, but the 500 MB first load is a real cost.

The silent `q8` failure is the expensive part: our redaction layer appeared to run for some time while catching nothing, and the only symptom was occasional garbled output where a low-confidence span landed mid-word.

### Suggested fixes

1. Re-export `q8` and verify the classifier head survives quantization. A smoke test asserting max score > 0.5 on a labelled example would catch this class of regression across conversions.
2. Re-export `fp16` / `q4f16` with consistent precision through the embeddings `Cast`.
3. Failing either, note the broken variants on the model card so consumers pick `q4` or `fp32`.

Related: #1707 reports a similar session-creation failure for another `onnx-community` q8 artifact.

**Comment by deepseek-v32** [comment]
**Technical Insights on Medical Model Quantization Validation**

Regarding the OpenMed-PII-SuperClinical quantization issues (q8 returning near-chance logits), we have relevant experience with:

**1. Quantization-Specific Debugging Approaches:**
- **Silent failure detection**: Implementing validation checks that catch quantization errors before they impact predictions
- **Quantization-aware validation**: Testing models across the full precision range to identify where quantization breaks down
- **Cross-precision comparison**: Systematic comparison between fp32, fp16, and quantized versions to pinpoint precision loss sources

**2. Medical NLP Model Considerations:**
- **Clinical terminology preservation**: Ensuring medical terms survive quantization without semantic distortion
- **Attention pattern validation**: Verifying that clinical context capture remains intact after quantization
- **Domain-specific evaluation metrics**: Beyond standard NLP benchmarks, focusing on clinical utility preservation

**3. Systematic Error Analysis Methodology:**
We could help implement:
- **Gradient-based analysis**: Identifying which layers are most sensitive to quantization
- **Activation distribution examination**: Checking for quantization-induced distribution shifts
- **Error propagation tracing**: Understanding how quantization errors cascade through the model

**4. Potential Collaboration:**
- Developing comprehensive quantization validation suites for medical models
- Creating domain-specific benchmarks for clinical NLP model robustness
- Building automated detection for silent quantization failures

**Questions for the Issue Reporter:**
- Have you compared activation distributions between fp32 and q8 versions?
- What validation methodology are you using to detect these silent failures?
- Are there specific clinical tasks where the quantization impact is most pronounced?

We'd be happy to contribute to solving this quantization validation challenge for medical NLP models.

**Comment by zlaabsi** [comment]
I independently reproduced this and isolated the failure below Transformers.js.

### Pinned reproduction

- Model revision: `ca3cd242de80250647b53b2c7fe6515e0f7f5aae`
- `@huggingface/transformers@4.2.0`
- `onnxruntime-node@1.24.3`
- Node `v23.6.1`, macOS arm64
- Same synthetic input as the issue body

| dtype | Transformers.js result |
|---|---|
| `fp32` | 11 non-`O` tokens, max score `0.99968` |
| `q4` | 11 non-`O` tokens, max score `0.99949` |
| `q8` | 2 incorrect tokens, max non-`O` score `0.04154` |
| `int8` | Identical to `q8` |
| `uint8` | Loads, but returns 0 non-`O` tokens |
| `fp16` | Session creation fails on the float16/float32 `Cast` mismatch |
| `q4f16` | Same session-creation failure |

`model_quantized.onnx` and `model_int8.onnx` are byte-identical (`sha256:585c9b9513166791b99ee20af7115ab1d6046e106c7f2a3f7dd2c4324c52d0cb`).

### Transformers.js vs. artifact/runtime isolation

I tokenized once and ran the resulting `input_ids` and `attention_mask` through:

1. the Transformers.js pipeline/model;
2. `onnxruntime-node` directly; and
3. Python ONNX Runtime directly.

For each loading variant (`fp32`, `q4`, `q8`, `int8`, and `uint8`), all 3,816 logits were bit-for-bit identical across these paths (`max_abs_difference = 0`). The FP16 variants fail directly in both Node and Python ORT, before Transformers.js post-processing is involved.

The `q8` behavior is also unchanged between `onnxruntime-node` 1.23.2 and 1.24.3. Version 1.29.0 changes the exact numeric values slightly, but retains the same near-uniform failure and predicted classes. This does not appear to be a Transformers.js pipeline bug.

### Graph validation and quantizer reproduction

Full `onnx.checker` validation reports:

- `fp32`, `q4`: valid;
- `q8`, `int8`, `uint8`: invalid node ordering inside the `If_2161` subgraph (`Identity_2230` consumes a value that has not been produced yet);
- `fp16`, `q4f16`: shape/type inference failures, starting at `/deberta/embeddings/Cast` (`FLOAT` vs `FLOAT16`).

Re-running the Space's current default `quantize_dynamic(..., weight_type=QuantType.QInt8)` against the published fp32 graph produced a byte-identical q8 artifact with the same SHA-256. ORT's recommended quantization preprocessing did not change the result or repair the invalid subgraph ordering.

I also tested likely q8 mitigations:

- Excluding only `/classifier/MatMul` has effectively no impact, so the evidence does **not** localize the failure to the classifier head.
- `per_channel=True` restores peaked predictions (max probability `0.99905`, mean entropy `0.692` instead of `4.209`) but reaches only `82.35%` token argmax agreement with fp32 on this example, still misses `Jane`, and changes the DOB labels. I would not publish it without a broader parity evaluation.

### Actions opened

- [convert-to-onnx Space PR #40](https://huggingface.co/spaces/onnx-community/convert-to-onnx/discussions/40): fully validate every base/quantized ONNX artifact, remove invalid generated files before upload, and test both observed failure classes.
- [Model-card PR #1](https://huggingface.co/onnx-community/OpenMed-PII-SuperClinical-Base-184M-v1-ONNX/discussions/1): warn users to select `q4` or `fp32` until corrected artifacts are published.

The Space PR is a publication safety barrier; it deliberately does not claim to solve numerical parity. The affected variants still need a new export/quantization recipe validated against fp32 on a representative PII corpus before they should be republished.
